### DISCUSSION TITLE
Add option to `noExplicitAny` which allows it be used in type argument `extends` clause

### DISCUSSION THREAD

**Discussion by juliusmarminge** [biomejs/biome#1663] [Rule suggestion]
I often find myself disabling the `any` lint rule for type parameters, would be nice to have an option to the lint rule to allow this usage without disabling the rule everytime manually:

![CleanShot 2024-01-25 at 11 04 29](https://github.com/biomejs/biome/assets/51714798/6dbbbc87-f2c4-4b51-81e3-d8c5cb205423)

**Comment by Conaclos** [comment]
Why don't you use `unknown`?

**Reply by juliusmarminge** [reply]
Because its not equivalent and i want to use any in this case

**Reply by Conaclos** [reply]
Could you elaborate on the difference? I would like to clearly understand the implication of enabling this.

**Reply by juliusmarminge** [reply]
Well let's look at this case for example, I'm saying that T extends any function (takes any arguments and can return anything). I am then serializing what the function returns, putting it in a cache. When reading from the cache, I want to deserialize it. With a strongly typed serializer it can look like this:

![CleanShot 2024-01-25 at 15 41 06](https://github.com/biomejs/biome/assets/51714798/010dd77c-c18f-4dd9-b247-0674e2ad802f)

Changing to unknown, I am now going to get a type-error further down the road:

![CleanShot 2024-01-25 at 15 41 27](https://github.com/biomejs/biome/assets/51714798/a162dd97-06d1-44b7-bc4c-2f24b25ec30f)

So let's fix that:

![CleanShot 2024-01-25 at 15 42 11](https://github.com/biomejs/biome/assets/51714798/67e8e764-f6d9-489e-b67e-69329844ea12)

aah crap, now we get new ones. Let's fix those too. Now it looks like the function is good right?

![CleanShot 2024-01-25 at 15 43 34](https://github.com/biomejs/biome/assets/51714798/fdceab1b-c452-4c91-a590-3c48460866fb)

Well now let's add some argument to the functions and we got problems again:

![CleanShot 2024-01-25 at 15 44 18](https://github.com/biomejs/biome/assets/51714798/25fa8d49-fba2-4300-8993-11f8674f1776)

So yea, the only way to get this to work is using `any`. And even if it would be possible in another way, you have to do a bunch of workarounds that often might end up in an `as any`...

![CleanShot 2024-01-25 at 15 45 03](https://github.com/biomejs/biome/assets/51714798/f8e34a9e-777b-4d49-a072-1f212f6b15ec)

**Reply by Conaclos** [reply]
Could you provide a TypeScript playground link of your first code (with `any`)?

Taking a closer look, I think it is a bad idea to allow this, because it is unsafe, it can lead to runtime errors.
For example:

```ts
function withUnstableCache<T extends (args: any[]) => any>(
	f: T,
	args: Parameters<T>,
	tags: CacheTags[],
) {
	f(""); f(0); f([]); // everything is accepted :o
}
```

I think you could solve the issue by using a [type parameter for the function parameters instead of the entire function](https://www.typescriptlang.org/play?#code/FAMwrgdgxgLglgewgAhAHgCrIKYA8bYQAmAzsgBQB01AhgE4DmJAXMjRAJ4DaAugJTIAvAD42nYeRARWGADRtGLZAAV6NALbYCdEpmECA3sGQnUEKrUV9jpqeQAMAgPRPkdSPE046dBHWAAvsjAoOTkNKwkMHRwEAwCImzyXABEKfwhoJCwiCgMmDj4hKQU1JT0TKx02DRESAA2HMiQANYQCADuELwJoq3tXRJSMvIVSqp0GlrYOnqGNiZ2ZWPWpmYOzq7uEJ7Y3r7+ASEMYRHIUTFxvUnIqenWIeDQ8EjIABZoAIKKhQTEZNVag0mv1Ot0eENpKVLJVkN8mNdQV1RopWPCSPM1ksYRiFutHMgXMgAOIIBBEACEyAA6nsGFo2MgoAh1AAHOD1Gb7PzIZjWI7AN6nSLRWLxISiGiUEis+pwGDkFLIFJ8ZIpN7Yer1BDIDp+epEe7AIA):

```ts
function withUnstableCache<Args extends readonly unknown[]>(
	f: (...args: Args) => unknown,
	args: Args,
	tags: CacheTags[],
) {
	// ...
}
```

**Reply by juliusmarminge** [reply]
https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwHYGcYEMBGAbAUwH0BjLEgCwLgF84AzKCEOAIlQIA8YB6MygqwDcAWABQoSLERwSUAlhgEAKugipaDJi1YxkYQmtTDx4kusxwYRuAF5Z8xSqMAKBDACeYAugBccAG0AXRoASlMxcS4peHMMeABhAEEEgAkAUSJlJIBxAGU7RHE4OCT-VixWcTosdFkLGFExT284BPIqZSwAc0KXFoIIejaUjKzcvNCAgGsCDyGrL0Hh5LTM7PygpqjOGLhaj1QSBjQSGGB1OAB3YBgKAFV43EJ2gQAeYrhlOC4lVAATOouAB0oKwUG6fn2qA8wVCdgAfNCPAAacQIlwQMDWfxIeiofzKIT7CFQgAK4KwIAISig6DeygRxOwkP8r06PWCtHhCE+cUs-Co-wASj5kHh4PYsFcsLcUE98MRBQQXJ8SgcjnAQWDSfDbEjeWISsb6vE4PJ0OLJfsZXKsdZgfjtcDwZDQk0TSV5HooBprOpgegCFBgFg8MAAF4qi1W91q2hoo3G+3oYEs9CJk1IdP+FNpnp1GiZuChZ1513oOORJPe5C+qxGYH-HzB0PhqNvJK2pT-N6in2oZRLBkI0cuZUisUSqs0CL8+D4gCMhQ1xxcWH8mBDqG6KLgOH8qGQIBwwb1SNY9AgEFY+zq87MDQYqAATCv0Ic1yncQxr4fj6eUDcoibA4OCt61KamBzk+FrLlK3bXLcDwKi8HQqoaJQ8DwJQAHoAPyfPi-hLsWFb+AEFSsHuL5BMWOaBKsYwbHkwJJEE1ThNW87mj4b4IbK8A3HcjyYM8BDshhnzYXhhFJsRz4vmRpIUQgV4QP4L40HRnwMQETHrBMbEcWIYQRMUwkoWJiqSW40k4XABFEQSinKaygRqX+oHgdp9EFhRBnjPkxmcRZyGidgNnoXZSYyY5cklAppGfORHnqZpvm6f5jGjIZwXsZxQA

**Reply by Conaclos** [reply]
What do you think of [this refactoring](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwHYGcYEMBGAbAUwH0BjLEgCwLgF84AzKCEOAIlQIA8YB6MygqwDcAWABQoSLERwSUAlhgEAKugipaDJi1YxkYQmtTDx4kusxwYRuAF5Z8xSqMAKBDACeYAugBccAG0AXRoASlMxcS4peHMMeABhAEEEgAkAUSJlJIBxAGU7RHE4OCT-VixWcTosdFkLGFExT284BPIqZSwAc0KXFoIIejaUjKzcvNCAgGsCDyGrL0Hh5LTM7PygpqjOGLhaj1QSBjQSGGB1OAB3YBgKAFV43EJ2gQAeYtKobrquJVQACZ1RwA9R4DwoVDTVAQK6oYIAGjgACUCPA-gRAXU0NDYahxAA+FwQMDWfxIeiofwuAB0dKw3z8Xx+oTsBJRaKE+0Z-iSjK52B+-lenR6wVorIQnziln4VABqPQyDw8HsWCuWFukMwz2IcoILk+JQORzgtPpjNZtnZUrEJXt9XicHkSpVhXVmvgJOsNMp5ppDJ++2xUJhcODzPQoSaDpK8j0UA01nUNPQBCgwCweGAAC8DS7lTBo0baAiS970DTBegy3b7Uhq-4K1WenUaLWSqF-c3A1HtnX48hE1YjDSAT505ns3m3kkNbcCAC3qiYASifqFT5C8WxDQIjL4JSAIzu9CHY4uLD+TAZ1DdJE4fyoZAgHDpq3s1j0CAQVgRg9mA0DCoAATKe55mhW5IMD+T4vm+UASmybA4Ayf61I6mD7kBLonmq87wDcdyPDq+AECKBq2iUPA8CUAB6AD8nyUv4x4dtyQqBBUrBIiBQTsY2gSrGMGx5DSSRBNU4SRGIB7Oj4YH4Z61y3A8TxkRRbifDR9FMXWLHASB7G9v4AQIN+ED+CBND8Z8gkBMJ6wTOJkm7tJEQqcR6kvB0lHabRcCMcxVKGcZPKBOZsFwAADDZAmtqZjnjPkLlScURFqaRPkCFpdY6YFeklAZbGfCZXE3jxMW2XW9lJaJqVuUAA)?

**Reply by Conaclos** [reply]
You could also use `never`:

```ts
function withUnstableCache<T extends (args: never[]) => any>(
	f: T,
	args: Parameters<T>,
	tags: CacheTags[],
) {
	// ...
}
```

See also [this comment](https://github.com/biomejs/biome/issues/1882#issuecomment-1959193145) for more details.
